src/v/kafka/server/group.cc

```diff
@@ -509,10 +526,186 @@ group::join_group_unknown_member(join_group_request&& r) {
+                            member,
+                            make_join_error(
+                              unknown_member_id,
+                              map_store_offset_error_code(result.error())));
+                          return;
+                      }
+                      // leader    -> member metadata
+                      // followers -> []
+                      std::vector<member_config> md;
+                      if (is_leader(new_member_id)) {
+                          md = member_metadata();
+                      }
+
+                      auto reply = join_group_response(
+                        error_code::none,
+                        generation(),
+                        protocol().value_or(kafka::protocol_name()),
+                        leader().value_or(kafka::member_id()),
+                        new_member_id,
+                        std::move(md));
+                      try_finish_joining_member(member, std::move(reply));
+                  });
+            return join_group_stages(std::move(f));
+        } else {
+            vlog(
+              _ctxlog.trace,
+              "static member {} joins in stable state with the protocol that "
+              "requires group protocol update, trying to trigger rebalance",
+              r.data.group_instance_id);
+            try_prepare_rebalance();
+        }
+        return join_group_stages(std::move(f));
+    }
+    case group_state::preparing_rebalance:
+        return join_group_stages(std::move(f));
+    case group_state::completing_rebalance:
+        vlog(
+          _ctxlog.trace,
+          "updating static member {} (member_id: {}) metadata",
+          member->group_instance_id().value(),
+          member->id());
+        try_prepare_rebalance();
+        return join_group_stages(std::move(f));
+    case group_state::empty:
+        [[fallthrough]];
+    case group_state::dead:
+        throw std::runtime_error(fmt::format(
+          "group was not supposed to be in {} state when the unknown static "
+          "member {} rejoins, group state: {}",
+          state(),
+          r.data.group_instance_id,
+          *this));
+    }
+}
+
+member_ptr group::replace_static_member(
+  const group_instance_id& instance_id,
+  const member_id& old_member_id,
+  const member_id& new_member_id) {
+    auto it = _members.find(old_member_id);
+    if (it == _members.end()) {
+        throw std::runtime_error(fmt::format(
+          "can not replace not existing member with id {}, group state: {}",
+          old_member_id,
+          *this));
+    }
+    auto member = it->second;
+    _members.erase(it);
+    member->replace_id(new_member_id);
+    _members.emplace(new_member_id, member);
 
-    // <kafka>Only return MEMBER_ID_REQUIRED error if joinGroupRequest version
-    // is >= 4 and groupInstanceId is configured to unknown.</kafka>
-    if (r.version >= api_version(4) && !r.data.group_instance_id) {
+    if (member->is_joining()) {
```

Comment:
why is it ok to remove this condition that is dependent on request version, and which is handled by upstream kafka? is the same condition handled in some other way?

Reply:
i think that `diff` is here misleading. This is entirely new code path that was introduced to handle members with `group.iinstance.id` assigned

Reply:
yeh, but i guess what i'm saying is that `if (r.version >= api_version(4) && !r.data.group_instance_id) {` is a specific condition in the protocol controlling behavior that clients expect. and it's not clear if that behavior is preserved and where its new location is at in the implementation.

Reply:
ahh i get it now. it's implied in add_new_dynamic_member and then the version check is relocated there!